dash/dash-renderer/src/actions/paths.js

```diff
@@ -46,7 +48,14 @@ export function computePaths(subTree, startingPath, oldPaths, events) {
                 const values = props(keys, id);
                 const keyStr = keys.join(',');
                 const paths = (objs[keyStr] = objs[keyStr] || []);
-                paths.push({values, path: concat(startingPath, itempath)});
+                const oldie = oldObjs[keyStr] || [];
+                const item = {values, path: concat(startingPath, itempath)};
+                const index = indexOf(item, oldie);
+                if (index === -1) {
+                    paths.push(item);
+                } else {
+                    objs[keyStr] = insert(index, item, paths);
+                }
```

Comment:
I can't quite put my finger on it, but I think there's a situation here where either the index you end up inserting at isn't what it should be anymore, or the item looks new - so gets put at the end instead of in the middle - but it's just a modification of the old item.

What I think we implicitly really want is for `objs[keystr]` to wind up sorted by path (ie in increasing value of the first item in the path that differs between any two elements). And if I'm reading this correctly, the only case in which this matters is when we saved some of the old paths [(line 35 above)](https://github.com/plotly/dash/pull/2388/files#diff-c151cd2dc301d2ecf998517c7dad59aebf33b90d530b725b889a37b11ffda958R35) and added more items with the same `keyStr` here.

So perhaps we could do something like save a record of which `keyStr`'s (`oldKeys`) we pre-populated above, then when we get here if we find a `keyStr` in that set we insert the new item in the list before the first item with a later path, if any; otherwise we push the new item on the end.

Reply:
I am not sure there is way currently to trigger the situation you are describing, the issue this fix is that the fix for pattern matching callback with component as props reevaluate paths of the parent of the prop and thus was updating the existing pmc indexes. Otherwise currently the subtree is updated entirely from the callback and evaluated properly.

Reply:
OK - I can't construct one either. Maybe partial props will expose something new... we'll see. Just keep this in mind as we go along in case any new bug reports pop up!